### PR TITLE
[move source language] Added spec id ~> code offset to compiled unit

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Added Nop's with optional label to IR
- The source map now has a map from nop label ~> offset
- The source language compiled unit maps from spec id ~> offset

## Test Plan

- cargo test

### CODE REVIEW

**Comment by wrwg** [comment]
> [move-prover] Move Prover test (`language/move-prover/cargo test`) on commit [f630831](https://github.com/libra/libra/commit/f630831570562d61fdaa54a79f720833c5e2427b) did **fail**. See [details](https://github.com/mimoo/mirai-bot/runs/520716143?check_suite_focus=true#step:7:1).

This will go away once #3037 is merged. This PR hangs in the queue for 5 hours now *_&*_()&*(_!

File: language/compiler/bytecode-source-map/src/source_map.rs [hunk 1]
```diff
@@ -191,6 +196,11 @@ impl<Location: Clone + Eq> FunctionSourceMap<Location> {
         };
     }
 
+    /// Record an the code offset for an Nop label
```
Comment (shazqadeer) [inline language/compiler/bytecode-source-map/src/source_map.rs]: Fix grammar in comment.

File: language/compiler/bytecode-source-map/src/source_map.rs [hunk 2]
```diff
@@ -191,6 +196,11 @@ impl<Location: Clone + Eq> FunctionSourceMap<Location> {
...
+    pub fn add_nop_mapping(&mut self, label: NopLabel, offset: CodeOffset) {
+        assert!(self.nops.insert(label, offset).is_none())
+    }
+
     // Not that it is important that locations be added in order.
```
Comment (shazqadeer) [inline language/compiler/bytecode-source-map/src/source_map.rs]: Not -> Note

File: language/move-lang/src/to_bytecode/translate.rs [hunk 3]
```diff
@@ -546,8 +602,8 @@ fn exp_(context: &mut Context, code: &mut IR::BytecodeBlock, e: H::Exp) {
     match e_ {
         E::Unreachable => panic!("ICE should not compile dead code"),
         E::UnresolvedError => panic!("ICE should not have reached compilation if there are errors"),
-        E::Spec(_) => (),
         E::Unit => (),
+        E::Spec(id) => code.push(sp(loc, B::Nop(Some(context.nop_label(id))))),
```
Comment (shazqadeer) [inline language/move-lang/src/to_bytecode/translate.rs]: Looks like we are pushing Some(NopLabel) every time.  So None will be used for some other purpose.  I am wondering what is the use case for Nop(None).

Comment (tnowacki) [reply language/move-lang/src/to_bytecode/translate.rs]: Its just for API design. The Nop in the IR might be used just as a normal Nop in some situations.

**Pull request closed without merging**